packages/core/diagnostic/src/diagnostic.js

```diff
@@ -0,0 +1,125 @@
+// @flow
+
+export type DiagnosticHighlightLocation = {|
+  line: number,
+  column: number
+|};
+
+export type DiagnosticSeverity = 'error' | 'warn' | 'info';
+
+export type DiagnosticCodeHighlight = {|
+  start: DiagnosticHighlightLocation,
+  end: DiagnosticHighlightLocation,
+  message?: string
+|};
+
+export type DiagnosticCodeFrame = {|
+  code: string,
+  codeHighlights: DiagnosticCodeHighlight | Array<DiagnosticCodeHighlight>
+|};
+
+// A Diagnostic is a style agnostic way of emitting errors, warnings and info
+// The reporter's are responsible for rendering the message, codeframes, hints, ...
+export type Diagnostic = {|
+  message: string,
+  origin: string, // Name of plugin or file that threw this error
+
+  // Asset metadata
+  filename?: string,
+  language?: string,
+
+  // Codeframe data
+  codeframe?: DiagnosticCodeFrame,
+
+  // Stacktrace for error, not really needed if there's a codeframe...
+  stack?: string,
+
+  // Hints to resolve issues faster
+  hints?: Array<string>
+|};
+
+// This type should represent all error formats Parcel can encounter...
+export type PrintableError = Error & {
+  fileName?: string,
```

Comment:
nit: `filePath: FilePath`

Reply:
This is part of the mozilla Error spec https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/fileName

Reply:
I think this is reason for us to choose something else like `filePath`. It looks like that `fileName` is the script file that raised the error itself [0], but this should be e.g. where a parse error occurred during parsing.


[0] https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/fileName

Reply:
PrintableError is actually the type for errors not created by parcel